platforms/astro.ts

```diff
@@ -0,0 +1,23 @@
+import type { APIContext, APIRoute } from 'astro'
+
+import type { WorkflowServeOptions, WorkflowContext } from "../src";
+import { serve as serveBase } from "../src";
+
+export function serve<TInitialPayload = unknown>(
+  routeFunction: (
+    apiContext: APIContext,
+    workflowContext: WorkflowContext<TInitialPayload>
+  ) => Promise<void>,
```

Comment:
Hi @hckhanh,

Thanks for the contribution! I would like to ask why we include the astro `APIContext` in the route function.

We didn't put any platform dependent objects in the route function so far. Is it needed in the case of Astro?

Reply:
@CahidArda  This is one of my implementations with the idea of the PR

```ts
import type { LinkDeletionInput } from '@/lib/types.ts'

import { qstash } from '@/lib/upstash/qstash.ts'
import { qstashReceiver } from '@/lib/upstash/qstash/receiver.ts'
import { serveWorkflow } from '@sshare/lib'
import { links } from '@sshare/schema'
import { APP_HOMEPAGE } from 'astro:env/client'
import { isNull, lte, or, sql } from 'drizzle-orm'
import { drizzle } from 'drizzle-orm/d1'
import { chunk } from 'es-toolkit'

export const { POST } = serveWorkflow(
  async (ctx, workflow) => {
    const { env } = ctx.locals.runtime

    // Get all links were accessed before 7 days from now or not access at all
    const results = await workflow.run('get-all-unused-links', async () =>
      drizzle(env.USER_DB)
        .select()
        .from(links)
        .where(
          or(
            isNull(links.lastAccessedAt),
            lte(
              links.lastAccessedAt,
              sql`CAST(strftime('%s', 'now', 'start of day', '-7 days') AS INTEGER)`
            )
          )
        )
    )

    if (!results.length) {
      return
    }

    await workflow.run('delete-links', async () => {
      const linksChunks = chunk(results, 38_000)
      return workflow.qstashClient.batchJSON<LinkDeletionInput[]>(
        linksChunks.map((linksChunk) => ({
          url: `${APP_HOMEPAGE}/api/upstash/qstash/workflows/delete-links`,
          body: linksChunk.map((link) => ({
            id: link.id,
            slug: link.slug,
            type: link.type,
            protect: link.passwordProtected,
          })),
        }))
      )
    })
  },
  { receiver: qstashReceiver.rawReceiver, qstashClient: qstash }
)
```

As you can see, I need the ctx object to call Cloudflare APIs. So It is much more convenient if we can pass the platform's context object to the workflow.

Reply:
> We didn't put any platform dependent objects

I think it's not a good point to isolate platform data with the route function because you know that the workflow needs to access everything in the platform. Otherwise, I have to put the route function inside the api routes of the platform. Is that the intended design of the Upstash workflow?

Reply:
We can pass the ctx.env to the workflow context using the [env option](https://upstash.com/docs/workflow/basics/serve#env), [like we do in cloudflare](https://github.com/upstash/workflow-js/blob/main/platforms/cloudflare.ts#L66).

But I agree that having the complete object is more convenient from a DX perspective. We can do it in Astro. If it works well we can update the serve methods of other platforms.

Reply:
I am glad that you agree with me.